pkg/controllermanager/controller/certificatesigningrequest/csr_autoapprove_control.go

```diff
@@ -64,75 +68,129 @@ type csrReconciler struct {
 }
 
 func (r *csrReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
-	csr := &certificatesv1beta1.CertificateSigningRequest{}
-	if err := r.gardenClient.Client().Get(ctx, request.NamespacedName, csr); err != nil {
+	var (
+		csrLogger = logger.NewFieldLogger(logger.Logger, "csr", request.Name)
+
+		cert               []byte
+		finalState         bool
+		req                []byte
+		usages             []certificatesv1.KeyUsage
+		extra              = make(map[string]authorizationv1.ExtraValue)
+		username           string
+		uid                string
+		groups             []string
+		updateConditionsFn func() error
+	)
+
+	csrV1 := &certificatesv1.CertificateSigningRequest{}
+	if err := r.gardenClient.Client().Get(ctx, request.NamespacedName, csrV1); err != nil {
 		if apierrors.IsNotFound(err) {
 			r.logger.Infof("Object %q is gone, stop reconciling: %v", request.Name, err)
 			return reconcile.Result{}, nil
 		}
-		r.logger.Infof("Unable to retrieve object %q from store: %v", request.Name, err)
-		return reconcile.Result{}, err
-	}
 
-	csrLogger := logger.NewFieldLogger(logger.Logger, "csr", csr.Name)
-
-	if len(csr.Status.Certificate) != 0 {
-		return reconcile.Result{}, nil
-	}
+		if !meta.IsNoMatchError(err) {
+			r.logger.Infof("Unable to retrieve object %q from store: %v", request.Name, err)
+			return reconcile.Result{}, err
+		}
 
-	for _, c := range csr.Status.Conditions {
-		if c.Type == certificatesv1beta1.CertificateApproved {
-			return reconcile.Result{}, nil
+		// fallback to v1beta1
```

Comment:
Always trying `v1` first, will cause quite some discovery calls (dynamic rest mapper refreshes rest mappings on `NoMatchErrors`).
When setting up the informer we already know which API version is available. Can you use that information and save the version we will want to use later on in the reconciler?